galley/pkg/config/analysis/analyzers/deprecation/deprecation.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deprecation
+
+import (
+	"istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+// FieldAnalyzer checks for deprecated Istio types and fields
+type FieldAnalyzer struct{}
+
+// Currently we don't have an Istio API that tells which Istio APIs are deprecated.
+// Run `find . -name "*.proto" -exec grep -i "deprecated=true" \{\} \; -print`
+// to see what is deprecated.  This analyzer is hand-crafted.
+
+// Metadata implements analyzer.Analyzer
+func (*FieldAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "deprecation.DeprecationAnalyzer",
+		Inputs: collection.Names{
+			metadata.IstioNetworkingV1Alpha3Virtualservices,
+			// TODO /mesh/v1alpha1/config.proto, /mesh/v1alpha1/proxy.proto
+		},
+	}
+}
+
+// Analyze implements analysis.Analyzer
+func (fa *FieldAnalyzer) Analyze(ctx analysis.Context) {
+	ctx.ForEach(metadata.IstioNetworkingV1Alpha3Virtualservices, func(r *resource.Entry) bool {
+		fa.analyzeVirtualService(r, ctx)
+		return true
+	})
+}
+
+func (*FieldAnalyzer) analyzeVirtualService(r *resource.Entry, ctx analysis.Context) {
+
+	vs := r.Item.(*v1alpha3.VirtualService)
+
+	for _, httpRoute := range vs.Http {
+		if httpRoute.WebsocketUpgrade {
+			ctx.Report(metadata.IstioNetworkingV1Alpha3Virtualservices,
+				msg.NewDeprecated(r, "HTTPRoute.websocket_upgrade ignored"))
+		}
+
+		if len(httpRoute.AppendHeaders) > 0 {
+			ctx.Report(metadata.IstioNetworkingV1Alpha3Virtualservices,
+				msg.NewDeprecated(r, "HTTPRoute.append_headers deprecated; use HTTPRoute.headers.request.add"))
```

Comment:
Using free text instead of multiple forms of the message means there will be a single code, `IST0004`.  That seems important -- we don't want the code to change if the suggestion changes from "use X" to "no longer needed."

I also initially wanted to normalize instead of writing ad-hoc text but I suspect there will be too many variations: "use this", "consult docs", "don't worry it's automatic", "only K8s", etc.